controller/internal/enforcer/proxy/enforcerproxy.go

```diff
@@ -169,27 +188,66 @@ func (s *ProxyInfo) SetLogLevel(level constants.LogLevel) error {
 
 // CleanUp sends a cleanup command to all the remotes forcing them to exit and clean their state.
 func (s *ProxyInfo) CleanUp() error {
+	var contextIDs []string
+	var synch sync.Mutex
+	var wg sync.WaitGroup
 
-	var allErrors string
+	s.Lock()
 
-	remotes := s.rpchdl.ContextList()
+	for contextID, _ := range s.remoteEnforcers {
```

Comment:
The list of remotes is already maintained in rpchdl .. adding one more map to maintain the list is possible to create problesm. Especially in the case of remote failures/restarts ... Try it out where you kill some remotes (or make them fail in a loop).